File: flutter/packages/measure_flutter/lib/src/events/custom_event_collector.dart
import 'dart:isolate';

import 'package:flutter/foundation.dart';
import 'package:measure_flutter/src/attribute_value.dart';
import 'package:measure_flutter/src/config/config_provider.dart';
import 'package:measure_flutter/src/events/custom_event_data.dart';
import 'package:measure_flutter/src/events/event_type.dart';
import 'package:measure_flutter/src/logger/logger.dart';
import 'package:measure_flutter/src/method_channel/signal_processor.dart';
import 'package:measure_flutter/src/time/time_provider.dart';

import '../logger/log_level.dart';

final class CustomEventCollector {
  final Logger logger;
  final SignalProcessor signalProcessor;
  final TimeProvider timeProvider;
  final ConfigProvider configProvider;
  bool _enabled = false;

  CustomEventCollector({
    required this.logger,
    required this.signalProcessor,
    required this.timeProvider,
    required this.configProvider,
  });

  void register() {
    _enabled = true;
  }

  void unregister() {
    _enabled = false;
  }

  void trackCustomEvent(
    String name,
    int? timestamp,
    Map<String, AttributeValue> attributes,
  ) {
    if (!_enabled) {
      return;
    }

    if (name.isEmpty) {
      logger.log(
        LogLevel.error,
        "Invalid event: name is empty",
      );
      return;
    }

    if (name.length > configProvider.maxEventNameLength) {
      logger.log(
        LogLevel.error,
        "Invalid event($name): name exceeds maximum length of ${configProvider.maxEventNameLength} characters",
      );
      return;
    }

    final RegExp customEventNameRegex =
        RegExp(configProvider.customEventNameRegex);
    if (!customEventNameRegex.hasMatch(name)) {
      logger.log(LogLevel.error, "Invalid event($name) format");
      return;
    }

    signalProcessor.trackEvent(
      data: CustomEventData(name: name),
      type: EventType.custom,
      timestamp: timestamp ?? timeProvider.now(),
      userDefinedAttrs: attributes,
      userTriggered: true,
      threadName: Isolate.current.debugName,
    );
  }

  @visibleForTesting
  bool isEnabled() {
    return _enabled;
  }
}

File: android/measure-android/measure/src/main/java/sh/measure/android/events/CustomEventCollector.kt
package sh.measure.android.events

import sh.measure.android.attributes.AttributeValue
import sh.measure.android.config.ConfigProvider
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.utils.TimeProvider
import java.util.concurrent.atomic.AtomicBoolean

internal class CustomEventCollector(
    private val logger: Logger,
    private val signalProcessor: SignalProcessor,
    private val timeProvider: TimeProvider,
    private val configProvider: ConfigProvider,
) {
    private val isEnabled = AtomicBoolean(false)
    private val customEventNameRegex by lazy { Regex(configProvider.customEventNameRegex) }

    fun register() {
        isEnabled.set(true)
    }

    fun unregister() {
        isEnabled.set(false)
    }

    fun trackEvent(name: String, attributes: Map<String, AttributeValue>, timestamp: Long?) {
        if (!isEnabled.get()) {
            return
        }
        if (!validateName(name)) {
            return
        }

        val data = CustomEventData(name)
        signalProcessor.track(
            timestamp = timestamp ?: timeProvider.now(),
            type = EventType.CUSTOM,
            data = data,
            userTriggered = true,
            userDefinedAttributes = attributes,
        )
    }

    private fun validateName(name: String): Boolean {
        if (name.isEmpty()) {
            logger.log(
                LogLevel.Error,
                "Invalid event: name is empty",
            )
            return false
        }

        if (name.length > configProvider.maxEventNameLength) {
            logger.log(
                LogLevel.Error,
                "Invalid event($name): name exceeds maximum length of ${configProvider.maxEventNameLength} characters",
            )
            return false
        }

        if (!name.matches(customEventNameRegex)) {
            return false
        }

        return true
    }
}
